### DISCUSSION TITLE
clipboard error in safari browser

### DISCUSSION THREAD

**Discussion by Wzhipeng** [zauberzeug/nicegui#2364] [Q&A]
### Question

```py
origin = await ui.run_javascript('location.origin')
share_link = f"{origin}/share"
js_code = f'''
    async function fetchData() {{
        const copyText = `{share_link}`;
        return copyText;
    }}
    const text = new ClipboardItem({{
      "text/plain": fetchData()
        .then(response => {{
            console.log(response);
            return response;
        }})
        .then(text => new Blob([text], {{ type: "text/plain" }}))
    }})
    navigator.clipboard.write([text]);
'''
ui.run_javascript(js_code, timeout=4)
```
Clicking the button in my app will run the above code and copy the content to the buffer. There is no problem in the Chrome browser, but an error is reported in Safari.
At the same time, I found that the copy of ui.code() is also invalid on Safari.
Does anyone know how to solve this problem? Too urgent

This is an error from safari:
<img width="1058" alt="image" src="https://github.com/zauberzeug/nicegui/assets/34640461/80d9f172-8455-4279-b2c8-2ffbfdc659ef">

​

**Comment by Wzhipeng** [comment]
@falkoschindler could you help me with this issue? Thanks

**Comment by falkoschindler** [comment]
Thanks for bringing this up, @Wzhipeng! I fixed it on #2377.

**Reply by Wzhipeng** [reply]
Thanks @falkoschindler  But if I need to process the data in onclick and copy the result of dynamic data to the buffer, how to achieve it?

**Reply by falkoschindler** [reply]
Oh, I see. You seem to have already implemented the dynamic data fetching in JavaScript. So you could wrap this logic in a JavaScript function `copyToClipboard`, add it via `add_head_html` and call it via an `onclick` prop:
```py
ui.add_head_html('''
<script>
    function copyToClipboard() {
        ...
    }
</script>
''')

ui.button('Click me').props('onclick="copyToClipboard()"')
```
